File: app/shared/src/main/java/com/topjohnwu/magisk/utils/DynamicClassLoader.java
package com.topjohnwu.magisk.utils;

import android.os.Process;

import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.util.Enumeration;

import dalvik.system.BaseDexClassLoader;

public class DynamicClassLoader extends BaseDexClassLoader {

    public DynamicClassLoader(File apk) {
        this(apk, DynamicClassLoader.class.getClassLoader());
    }

    public DynamicClassLoader(File apk, ClassLoader parent) {
        // Set optimizedDirectory to null for RootService to bypass DexFile's security checks
        super(apk.getPath(), Process.myUid() == 0 ? null : apk.getParentFile(), null, parent);
    }

    @Override
    protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
        // First check if already loaded
        Class<?> cls = findLoadedClass(name);
        if (cls != null)
            return cls;

        try {
            // Then check boot classpath
            return getSystemClassLoader().loadClass(name);
        } catch (ClassNotFoundException ignored) {
            try {
                // Next try current dex
                return findClass(name);
            } catch (ClassNotFoundException fromSuper) {
                try {
                    // Finally try parent
                    return getParent().loadClass(name);
                } catch (ClassNotFoundException e) {
                    throw fromSuper;
                }
            }
        }
    }

    @Override
    public URL getResource(String name) {
        URL resource = getSystemClassLoader().getResource(name);
        if (resource != null)
            return resource;
        resource = findResource(name);
        if (resource != null)
            return resource;
        resource = getParent().getResource(name);
        return resource;
    }

    @Override
    public Enumeration<URL> getResources(String name) throws IOException {
        return new CompoundEnumeration<>(getSystemClassLoader().getResources(name),
                findResources(name), getParent().getResources(name));
    }
}

File: app/src/main/java/com/topjohnwu/magisk/utils/DynamicClassLoader.kt
package com.topjohnwu.magisk.utils

import dalvik.system.DexClassLoader
import java.io.File
import java.io.IOException
import java.net.URL
import java.util.*

@Suppress("FunctionName")
inline fun <reified T> T.DynamicClassLoader(apk: File) = DynamicClassLoader(apk, T::class.java.classLoader)

class DynamicClassLoader(apk: File, parent: ClassLoader?)
    : DexClassLoader(apk.path, apk.parent, null, parent) {

    private val base by lazy { Any::class.java.classLoader!! }

    @Throws(ClassNotFoundException::class)
    override fun loadClass(name: String, resolve: Boolean) : Class<*>
        = findLoadedClass(name) ?: runCatching {
            base.loadClass(name)
        }.getOrElse {
            runCatching {
                findClass(name)
            }.getOrElse { err ->
                runCatching {
                    parent.loadClass(name)
                }.getOrElse { throw err }
            }
        }

    override fun getResource(name: String) = base.getResource(name)
            ?: findResource(name)
            ?: parent?.getResource(name)

    @Throws(IOException::class)
    override fun getResources(name: String): Enumeration<URL> {
        val resources = mutableListOf(
                base.getResources(name),
                findResources(name), parent.getResources(name))
        return object : Enumeration<URL> {
            override fun hasMoreElements(): Boolean {
                while (true) {
                    if (resources.isEmpty())
                        return false
                    if (!resources[0].hasMoreElements()) {
                        resources.removeAt(0)
                    } else {
                        return true
                    }
                }
            }

            override fun nextElement(): URL {
                if (!hasMoreElements())
                    throw NoSuchElementException()
                return resources[0].nextElement()
            }
        }
    }

}
